src/Server/TCPHandler.cpp

```diff
@@ -760,6 +760,13 @@ void TCPHandler::processTablesStatusRequest()
     request.read(*in, client_tcp_protocol_version);
 
     ContextPtr context_to_resolve_table_names = (session && session->sessionContext()) ? session->sessionContext() : server.context();
+    if (is_interserver_mode && !default_database.empty())
+    {
+        // in interserver mode, the session doesn't exist.
+        ContextMutablePtr interserver_context = Context::createCopy(context_to_resolve_table_names);
+        interserver_context->setCurrentDatabase(default_database);
+        context_to_resolve_table_names = interserver_context;
```

Comment:
Does separate context really required here, maybe using server's context is enough? (and pass `default_database` if `table_name.database.empty()`)
@tavplubix what do you think?

Reply:
It does not work because later we try to use session context (not query context) and end up using global context:
https://github.com/ClickHouse/ClickHouse/blob/90fa4863f9beabe094474e4d8516ddc1b0eac518/src/Server/TCPHandler.cpp#L762

(I don't understand what this code does, but maybe we should throw LOGICAL_ERROR if we don't have some local (session or query) context here. It does not make sense to use global one)

Reply:
>I don't like passing default_database here because it's ad-hoc trick, as well as I don't like creating separate context here for the same reason

Same here.

> But it does not work for some reason

It does not work, because it is too late, `Protocol::Client::Query` (`receiveQuery`) goes after `Protocol::Client::TablesStatusRequest` (`processTablesStatusRequest`)